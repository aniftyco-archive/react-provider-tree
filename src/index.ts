import { Fragment, createElement } from 'react';

export type Provider = [any, Record<string, any>];

const createProviderTree = (providers: Provider[], index: number = 0, children: any) => {
  const [Provider, props = {}] = providers[index];

  if (index === providers.length - 1) {
    return createElement(Provider, { ...props, children });
  }

  return createElement(Provider, props, createProviderTree(providers, ++index, children));
};

export const createProviderTreeFromList = (...providers: Provider[]) => {
  return ({ children }: { children?: any }) => {
    return createProviderTree(providers, 0, children);
  };
};
