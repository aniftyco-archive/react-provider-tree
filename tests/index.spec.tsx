import { render } from '@testing-library/react';
import React, { FunctionComponent } from 'react';
import { createProviderTreeFromList } from '../src/';

describe('react-provider-tree', () => {
  const App = () => <div>Application</div>;
  const MockProvider: FunctionComponent = ({ children, ...props }) => <div {...props}>{children}</div>;

  it('should render the tree of providers', () => {
    const Provider = createProviderTreeFromList(
      [MockProvider, { name: 'provider-a' }],
      [MockProvider, { name: 'provider-b' }],
      [MockProvider, { name: 'provider-c' }]
    );
    const { container } = render(
      <Provider>
        <App />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
