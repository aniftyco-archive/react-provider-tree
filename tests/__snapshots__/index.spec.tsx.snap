// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-provider-tree should render the tree of providers 1`] = `
<div>
  <div
    name="provider-a"
  >
    <div
      name="provider-b"
    >
      <div
        name="provider-c"
      >
        <div>
          Application
        </div>
      </div>
    </div>
  </div>
</div>
`;
